arbor/backends/gpu/stack.hpp

```diff
@@ -67,41 +81,67 @@ class stack {
         storage_(create_storage(capacity)), gpu_context_(gpu_ctx) {}
 
     ~stack() {
-        storage_.synchronize();
-        if (storage_->data) {
-            allocator<value_type>().deallocate(storage_->data, storage_->capacity);
+        memory::cuda_sync();
+        auto st = get_storage_copy();
+        if (st.data) {
+            allocator<value_type>().deallocate(st.data, st.capacity);
         }
     }
 
-    // Perform any required synchronization if concurrent host-side access is not supported.
-    // (Correctness still requires that GPU operations on this stack are complete.)
-    void host_access() const {
-        gpu_context_->synchronize_for_managed_access();
+    std::vector<T> get_data() const {
+        if (capacity() != 0u) {
+
+            auto num = std::min(host_copy_.stores, host_copy_.capacity);
+            std::vector<T> buf(num);
+            auto bytes = num*sizeof(T);
+            memory::cuda_memcpy_d2h(buf.data(), host_copy_.data, bytes);
+            
+            return buf;
+        }
+        else {
+            return std::vector<T> {};
+        }
+
+    }
+
+    storage_type get_storage_copy() const {
+        storage_type st;
+        memory::cuda_memcpy_d2h(&st, storage_, sizeof(storage_type));
+
+        return st;
+    }
+
+
+    void refresh_host_copy() {
```

Comment:
You can replace `refresh_host_copy` and `get_storage_copy` with a single call `update_host`.